CRM/Financeextras/BAO/CreditNote.php

```diff
@@ -8,19 +7,79 @@ class CRM_Financeextras_BAO_CreditNote extends CRM_Financeextras_DAO_CreditNote
    *
    * @param array $params key-value pairs
    * @return CRM_Financeextras_DAO_CreditNote|NULL
+   */
+  public static function create($params) {
+    $className = 'CRM_Financeextras_DAO_CreditNote';
+    $entityName = 'CreditNote';
+    $hook = empty($params['id']) ? 'create' : 'edit';
+
+    CRM_Utils_Hook::pre($hook, $entityName, CRM_Utils_Array::value('id', $params), $params);
+    $instance = new $className();
+    $instance->copyValues($params);
+    $instance->save();
+    CRM_Utils_Hook::post($hook, $entityName, $instance->id, $instance);
+
+    return $instance;
+  }
+
+  /**
+   * Computes the credit note line item total.
    *
-   * public static function create($params) {
-   * $className = 'CRM_Financeextras_DAO_CreditNote';
-   * $entityName = 'CreditNote';
-   * $hook = empty($params['id']) ? 'create' : 'edit';
+   * @param array $items
+   *   Array of credit note line items.
    *
-   * CRM_Utils_Hook::pre($hook, $entityName, CRM_Utils_Array::value('id', $params), $params);
-   * $instance = new $className();
-   * $instance->copyValues($params);
-   * $instance->save();
-   * CRM_Utils_Hook::post($hook, $entityName, $instance->id, $instance);
+   * @return array
+   *   ['totalAfterTax' => <value>, 'totalBeforeTax' => <value>]
+   */
+  public static function computeTotal(array $items) {
+    $totalBeforeTax = round(array_reduce($items, fn ($a, $b) => $a + self::getSubTotal($b), 0), 2);
+    $totalAfterTax = round(array_reduce($items,
+      fn ($a, $b) => $a + ($b['tax_amount'] ?? (($b['tax_rate'] * self::getSubTotal($b)) / 100)),
+      0
+    ) + $totalBeforeTax, 2);
+
+    return [
+      'taxRates' => self::computeTaxRates($items),
+      'totalAfterTax' => $totalAfterTax,
+      'totalBeforeTax' => $totalBeforeTax,
+    ];
+  }
+
+  /**
+   * Computes the sub total of a single line item.
    *
-   * return $instance;
-   * } */
+   * @param array $item
+   *   Single credit note line item.
+   *
+   * @return int
+   *   The line item subtotal.
+   */
+  public static function getSubTotal(array $item) {
```

Comment:
getSubTotal -> getLineItemSubTotal


also it should be `private` given it is not used anywhere outside the class